File: src/main/java/io/salad109/conjunctionapi/conjunction/PairReduction.java
package io.salad109.conjunctionapi.conjunction;

import io.salad109.conjunctionapi.satellite.Satellite;

public class PairReduction {

    private PairReduction() {
    }

    /**
     * Determines if two satellites could possibly collide.
     * Applies orbital geometry filters with mathematical certainty.
     */
    public static boolean canCollide(Satellite a, Satellite b, double toleranceKm) {
        // Apply filters starting with computationally cheapest
        return altitudeShellsOverlap(a, b, toleranceKm) &&
                neitherAreDebris(a, b) &&
                orbitalPlanesIntersect(a, b, toleranceKm) &&
                areNotDocked(a, b);
    }

    private static boolean altitudeShellsOverlap(Satellite a, Satellite b, double toleranceKm) {
        double perigeeA = a.getPerigeeKm();
        double apogeeA = a.getApogeeKm();
        double perigeeB = b.getPerigeeKm();
        double apogeeB = b.getApogeeKm();
        return !(apogeeA + toleranceKm < perigeeB || apogeeB + toleranceKm < perigeeA);
    }

    private static boolean neitherAreDebris(Satellite a, Satellite b) {
        return !"DEBRIS".equals(a.getObjectType()) && !"DEBRIS".equals(b.getObjectType());
    }

    private static boolean orbitalPlanesIntersect(Satellite a, Satellite b, double toleranceKm) {
        double iA = Math.toRadians(a.getInclination());
        double iB = Math.toRadians(b.getInclination());
        double raanA = Math.toRadians(a.getRaan());
        double raanB = Math.toRadians(b.getRaan());
        double omegaA = Math.toRadians(a.getArgPerigee());
        double omegaB = Math.toRadians(b.getArgPerigee());
        double eA = a.getEccentricity();
        double eB = b.getEccentricity();
        double aA = a.getSemiMajorAxisKm();
        double aB = b.getSemiMajorAxisKm();

        double deltaRaan = raanA - raanB;

        // Coplanar orbits can intersect anywhere
        double relativeInclination = computeRelativeInclination(iA, iB, deltaRaan);
        if (relativeInclination < Math.toRadians(0.1)) {
            return true;
        }

        // Find where each orbit crosses the intersection line between orbital planes
        double alphaA = computeAlphaA(iA, iB, deltaRaan);
        double alphaB = computeAlphaB(iA, iB, deltaRaan);

        // Convert to true anomaly
        double nuA1 = normalizeAngle(alphaA - omegaA);
        double nuA2 = normalizeAngle(alphaA + Math.PI - omegaA);
        double nuB1 = normalizeAngle(alphaB - omegaB);
        double nuB2 = normalizeAngle(alphaB + Math.PI - omegaB);

        // Compute orbital radius at crossing points
        double rA1 = orbitalRadius(aA, eA, nuA1);
        double rA2 = orbitalRadius(aA, eA, nuA2);
        double rB1 = orbitalRadius(aB, eB, nuB1);
        double rB2 = orbitalRadius(aB, eB, nuB2);

        // If radii match at either crossing point, orbits can intersect
        // Check all four combinations
        double diff1 = Math.abs(rA1 - rB1);
        double diff2 = Math.abs(rA1 - rB2);
        double diff3 = Math.abs(rA2 - rB1);
        double diff4 = Math.abs(rA2 - rB2);

        double minDiff = Math.min(Math.min(diff1, diff2), Math.min(diff3, diff4));
        return minDiff <= toleranceKm;
    }

    private static double computeRelativeInclination(double iA, double iB, double deltaRaan) {
        double cosRelInc = Math.cos(iA) * Math.cos(iB)
                + Math.sin(iA) * Math.sin(iB) * Math.cos(deltaRaan);
        return Math.acos(Math.clamp(cosRelInc, -1, 1));
    }

    private static double computeAlphaA(double iA, double iB, double deltaRaan) {
        double sinIB = Math.sin(iB);
        double cosIB = Math.cos(iB);
        double sinIA = Math.sin(iA);
        double cosIA = Math.cos(iA);
        double sinDeltaRaan = Math.sin(deltaRaan);
        double cosDeltaRaan = Math.cos(deltaRaan);

        double y = sinIB * sinDeltaRaan;
        double x = sinIA * cosIB - cosIA * sinIB * cosDeltaRaan;

        return Math.atan2(y, x);
    }

    private static double computeAlphaB(double iA, double iB, double deltaRaan) {
        double sinIB = Math.sin(iB);
        double cosIB = Math.cos(iB);
        double sinIA = Math.sin(iA);
        double cosIA = Math.cos(iA);
        double sinDeltaRaan = Math.sin(deltaRaan);
        double cosDeltaRaan = Math.cos(deltaRaan);

        double y = -sinIA * sinDeltaRaan;
        double x = sinIB * cosIA - cosIB * sinIA * cosDeltaRaan;

        return Math.atan2(y, x);
    }

    private static double orbitalRadius(double semiMajorAxis, double eccentricity, double trueAnomaly) {
        double p = semiMajorAxis * (1 - eccentricity * eccentricity);
        return p / (1 + eccentricity * Math.cos(trueAnomaly));
    }

    private static double normalizeAngle(double angle) {
        while (angle > Math.PI) angle -= 2 * Math.PI;
        while (angle < -Math.PI) angle += 2 * Math.PI;
        return angle;
    }

    private static boolean areNotDocked(Satellite a, Satellite b) {
        return !(a.getEpoch() != null && a.getEpoch().equals(b.getEpoch()) &&
                a.getMeanMotion() != null && a.getMeanMotion().equals(b.getMeanMotion()) &&
                a.getEccentricity() != null && a.getEccentricity().equals(b.getEccentricity()) &&
                a.getInclination() != null && a.getInclination().equals(b.getInclination()) &&
                a.getRaan() != null && a.getRaan().equals(b.getRaan()) &&
                a.getArgPerigee() != null && a.getArgPerigee().equals(b.getArgPerigee()) &&
                a.getMeanAnomaly() != null && a.getMeanAnomaly().equals(b.getMeanAnomaly()));
    }
}

File: src/main/c/pair_reduction.c
#define _GNU_SOURCE
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>
#include <math.h>
#include <omp.h>

#ifndef M_PI
#define M_PI 3.14159265358979323846
#endif

typedef struct {
    int satellite_a_idx;
    int satellite_b_idx;
} SatellitePair;

static inline double clamp(double value, double min, double max) {
    if (value < min) return min;
    if (value > max) return max;
    return value;
}

static inline double normalize_angle(double angle) {
    while (angle > M_PI) angle -= 2.0 * M_PI;
    while (angle < -M_PI) angle += 2.0 * M_PI;
    return angle;
}

static inline double orbital_radius(double semi_major_axis, double eccentricity, double true_anomaly) {
    double p = semi_major_axis * (1.0 - eccentricity * eccentricity);
    return p / (1.0 + eccentricity * cos(true_anomaly));
}

static inline double compute_relative_inclination(double i_a, double i_b, double delta_raan) {
    double cos_rel_inc = cos(i_a) * cos(i_b) + sin(i_a) * sin(i_b) * cos(delta_raan);
    return acos(clamp(cos_rel_inc, -1.0, 1.0));
}

static inline double compute_alpha_a(double i_a, double i_b, double delta_raan) {
    double sin_i_b = sin(i_b);
    double cos_i_b = cos(i_b);
    double sin_i_a = sin(i_a);
    double cos_i_a = cos(i_a);
    double sin_delta_raan = sin(delta_raan);
    double cos_delta_raan = cos(delta_raan);

    double y = sin_i_b * sin_delta_raan;
    double x = sin_i_a * cos_i_b - cos_i_a * sin_i_b * cos_delta_raan;
    return atan2(y, x);
}

static inline double compute_alpha_b(double i_a, double i_b, double delta_raan) {
    double sin_i_b = sin(i_b);
    double cos_i_b = cos(i_b);
    double sin_i_a = sin(i_a);
    double cos_i_a = cos(i_a);
    double sin_delta_raan = sin(delta_raan);
    double cos_delta_raan = cos(delta_raan);

    double y = -sin_i_a * sin_delta_raan;
    double x = sin_i_b * cos_i_a - cos_i_b * sin_i_a * cos_delta_raan;
    return atan2(y, x);
}

static bool orbital_planes_miss(
    double inclination_a, double raan_a_deg, double arg_perigee_a,
    double e_a, double a_a,
    double inclination_b, double raan_b_deg, double arg_perigee_b,
    double e_b, double a_b,
    double tolerance_km)
{
    double i_a = inclination_a * (M_PI / 180.0);
    double i_b = inclination_b * (M_PI / 180.0);
    double omega_a = arg_perigee_a * (M_PI / 180.0);
    double omega_b = arg_perigee_b * (M_PI / 180.0);
    double delta_raan = (raan_a_deg - raan_b_deg) * (M_PI / 180.0);

    double relative_inclination = compute_relative_inclination(i_a, i_b, delta_raan);
    if (relative_inclination < (0.1 * (M_PI / 180.0))) {
        return false;
    }

    double alpha_a = compute_alpha_a(i_a, i_b, delta_raan);
    double alpha_b = compute_alpha_b(i_a, i_b, delta_raan);

    double nu_a1 = normalize_angle(alpha_a - omega_a);
    double nu_a2 = normalize_angle(alpha_a + M_PI - omega_a);
    double nu_b1 = normalize_angle(alpha_b - omega_b);
    double nu_b2 = normalize_angle(alpha_b + M_PI - omega_b);

    double r_a1 = orbital_radius(a_a, e_a, nu_a1);
    double r_a2 = orbital_radius(a_a, e_a, nu_a2);
    double r_b1 = orbital_radius(a_b, e_b, nu_b1);
    double r_b2 = orbital_radius(a_b, e_b, nu_b2);

    double diff1 = fabs(r_a1 - r_b1);
    double diff2 = fabs(r_a1 - r_b2);
    double diff3 = fabs(r_a2 - r_b1);
    double diff4 = fabs(r_a2 - r_b2);

    double min_diff = fmin(fmin(diff1, diff2), fmin(diff3, diff4));
    return min_diff > tolerance_km;
}

int filter_satellite_pairs(
    int n,
    const double* perigees,
    const double* apogees,
    const int* is_debris,
    const double* inclinations,
    const double* raans,
    const double* arg_perigees,
    const double* eccentricities,
    const double* semi_major_axes,
    double tolerance_km,
    SatellitePair* out_pairs)
{
    // physical cores only
    int num_threads = omp_get_num_procs() / 2;
    if (num_threads < 1) num_threads = 1;
    omp_set_num_threads(num_threads);

    // over-allocate double the survival rate
    long total_pairs = (long) n * (n - 1) / 2;
    int per_thread_cap = (int) (total_pairs / num_threads / 10) + 1024;

    int* thread_counts = (int*) calloc(num_threads, sizeof(int));
    SatellitePair** thread_buffers = (SatellitePair**) malloc(num_threads * sizeof(SatellitePair*));
    for (int t = 0; t < num_threads; t++) {
        thread_buffers[t] = (SatellitePair*) malloc(per_thread_cap * sizeof(SatellitePair));
    }

    #pragma omp parallel
    {
        int tid = omp_get_thread_num();
        SatellitePair* local_buf = thread_buffers[tid];
        int local_count = 0;
        int local_cap = per_thread_cap;

        #pragma omp for schedule(dynamic, 64) nowait
        for (int i = 0; i < n; i++) {
            double perigee_a = perigees[i];
            double apogee_a = apogees[i];
            int debris_a = is_debris[i];

            for (int j = i + 1; j < n; j++) {
                if ((apogee_a + tolerance_km < perigees[j])
                        || (apogees[j] + tolerance_km < perigee_a)
                        || (debris_a && is_debris[j])
                        || orbital_planes_miss(
                            inclinations[i], raans[i], arg_perigees[i],
                            eccentricities[i], semi_major_axes[i],
                            inclinations[j], raans[j], arg_perigees[j],
                            eccentricities[j], semi_major_axes[j],
                            tolerance_km)) {
                    continue;
                }

                if (local_count >= local_cap) {
                    local_cap *= 2;
                    local_buf = (SatellitePair*) realloc(local_buf, local_cap * sizeof(SatellitePair));
                    thread_buffers[tid] = local_buf;
                }

                local_buf[local_count].satellite_a_idx = i;
                local_buf[local_count].satellite_b_idx = j;
                local_count++;
            }
        }

        thread_counts[tid] = local_count;
    }

    // merge
    int pair_count = 0;
    for (int t = 0; t < num_threads; t++) {
        memcpy(out_pairs + pair_count, thread_buffers[t],
               thread_counts[t] * sizeof(SatellitePair));
        pair_count += thread_counts[t];
        free(thread_buffers[t]);
    }
    free(thread_buffers);
    free(thread_counts);

    return pair_count;
}
